Add Codable conformances for ShapedArray, TensorShape, and Tensor.

dan-zheng opened the pull request:
Encoding/decoding `Tensor` crashes for some reason. `ShapedArray` and `TensorShape` work fine.

dan-zheng committed:
[WIP] Add Codable conformances for ShapedArray, TensorShape, and Tensor.

Encoding/decoding Tensors crashes for some reason.

dan-zheng commented on stdlib/public/TensorFlow/ShapedArray.swift:
```diff
@@ -578,6 +578,27 @@ extension ShapedArray : CustomReflectable {
 
+/// Codable conformance.
+extension ShapedArray : Codable where Scalar : Codable {
+  private enum CodingKeys: String, CodingKey {
```
Currently, almost all `CodingKeys` types in stdlib SDKs conform to `Int`, not `String`:
```console
$ grep -nr "enum CodingKeys"
public/SDK/MediaPlayer/MPMusicPlayerPlayParameters.swift:64:    internal enum CodingKeys: String, CodingKey {
public/SDK/Foundation/URLComponents.swift:476:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/PersonNameComponents.swift:150:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/IndexSet.swift:859:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/IndexPath.swift:798:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/Calendar.swift:1132:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/DateComponents.swift:356:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/URL.swift:1221:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/TimeZone.swift:281:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/CharacterSet.swift:811:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/Measurement.swift:282:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/Locale.swift:485:    private enum CodingKeys : Int, CodingKey {
public/SDK/Foundation/Decimal.swift:473:    private enum CodingKeys : Int, CodingKey {
```
Here's what the difference looks like for `ShapedArray`:
```swift
// enum CodingKeys : String
// {"shape":[2,3],"scalars":[1,2,3,4,5,6]}
// enum CodingKeys : Int
// {0:[2,3],1:[1,2,3,4,5,6]}
```

rxwei commented on test/TensorFlowRuntime/codable.swift:
```diff
@@ -0,0 +1,54 @@
+}
+
+// NOTE: This test currently crashes with:
+// Compiler bug -- Tensor op builtin __tfop_Shape,$in,T,out_type cannot be lowered to LLVM IR
```
Try making tensor implementations `@inline(__always)`.

dan-zheng replied:
That didn't fix the crash, unfortunately.
```swift
// codable.swift
import TensorFlow
import Foundation

func testTensor() throws {
  let tensor = Tensor<Int32>(shape: [2, 3], scalars: Array(1...6))
  let encoder = JSONEncoder()
  let data = try encoder.encode(tensor)
}
try testTensor()
```
```console
$ swift -Xllvm -tf-dump-intermediates -O codable.swift
```

rxwei replied:
It is the case that partitioning won't work on throwing functions yet?

dan-zheng committed:
Merge branch 'tensorflow' into tensor-codable

dan-zheng committed:
Use `singleValueContainer` in Codable implementation for Tensor.

dan-zheng committed:
Fix warnings.

Fix the following warning:
"variable 'container' was never mutated; consider changing to 'let' constant"

dan-zheng committed:
Fix codable test.

dan-zheng committed:
Merge branch 'tensorflow' into tensor-codable

dan-zheng committed:
Work around GPE shortcoming in codable test.

`expectEqual` cannot be called with `Tensor` arguments because it is not
@inlinable and defined in another module. Calling it with `Tensor` arguments
results in a crash:

!!! Compiler bug -- Tensor op builtin __tfop_Equal,$in,$in,T cannot be lowered to LLVM IR !!!

rxwei commented on stdlib/public/TensorFlow/ShapedArray.swift:
```diff
@@ -889,3 +910,24 @@ extension ShapedArraySlice : CustomReflectable {
   }
 }
+
+/// Codable conformance.
```
Triple-slash doc comments on extensions are not picked up by any doc generator or Xcode, are they?

dan-zheng replied:
I'm sure they're not picked up by Jazzy, not sure about Xcode.
It would be better to use `// MARK`, probably. There are a bunch of triple slash comments in these files that should be changed to `// MARK`.

rxwei commented on test/TensorFlowRuntime/codable.swift:
```diff
@@ -0,0 +1,64 @@
+private func _testRoundTrip<T>(
+  of value: T, expectedJSON json: Data? = nil
+) where T : Codable, T : Equatable {
+  var payload: Data! = nil
```
I don't understand why this has to be an IUO.

If you just write
```swift
var payload: Data
```
... without initializing it, the `do-catch` block is guaranteed to initialize it anyway (by definitive initialization). I think it should compile.

rxwei replied:
Ok, I assumed that the `catch` block was trapping. Turns out it's `expectUnreachable` which doesn't return never. Then it makes sense to not compile.

dan-zheng committed:
Address comments by @rxwei.

- Add @inlinable to `encode` and `decode` for `TensorShape`.
- Clarify that `_testRoundTrip` is adapted from test/stdlib/TestJSONEncoder.swift.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
Apparently this is ready to land. Thanks Dan!
